syntax = "proto3";

package schedule;

// ============ СЕРВИС ============

service ScheduleService {
    // Просмотр расписания
    rpc GetScheduleForGroup(GroupScheduleRequest) returns (ScheduleResponse);
    rpc GetScheduleForTeacher(TeacherScheduleRequest) returns (ScheduleResponse);
    rpc GetScheduleForClassroom(ClassroomScheduleRequest) returns (ScheduleResponse);
    rpc GetScheduleForDay(DayScheduleRequest) returns (ScheduleResponse);
    rpc GetLesson(GetLessonRequest) returns (LessonResponse);
    
    // CRUD расписания
    rpc CreateLesson(CreateLessonRequest) returns (LessonResponse);
    rpc UpdateLesson(UpdateLessonRequest) returns (LessonResponse);
    rpc DeleteLesson(DeleteLessonRequest) returns (DeleteResponse);
    rpc BulkCreateLessons(BulkCreateRequest) returns (BulkCreateResponse);
    
    // Активация расписания (из ms-agent)
    rpc ActivateSchedule(ActivateRequest) returns (ActivateResponse);
    rpc DeactivateSchedule(DeactivateRequest) returns (DeactivateResponse);
    
    // Конфликты
    rpc CheckConflicts(ConflictCheckRequest) returns (ConflictResponse);
    rpc GetConflicts(GetConflictsRequest) returns (ConflictsListResponse);
    rpc ResolveConflict(ResolveConflictRequest) returns (ResolveResponse);
    
    // История
    rpc GetHistory(HistoryRequest) returns (HistoryResponse);
    rpc GetLessonHistory(LessonHistoryRequest) returns (HistoryResponse);
    
    // Снимки (snapshots)
    rpc CreateSnapshot(SnapshotRequest) returns (SnapshotResponse);
    rpc ListSnapshots(ListSnapshotsRequest) returns (SnapshotsListResponse);
    rpc RestoreSnapshot(RestoreSnapshotRequest) returns (RestoreResponse);
    rpc CompareSnapshots(CompareRequest) returns (CompareResponse);
    
    // Экспорт
    rpc ExportToExcel(ExportRequest) returns (ExportResponse);
    rpc ExportToPDF(ExportRequest) returns (ExportResponse);
    rpc ExportToICal(ExportRequest) returns (ExportResponse);
    
    // Поиск
    rpc SearchSchedule(SearchRequest) returns (ScheduleResponse);
    
    // Статистика
    rpc GetStatistics(StatisticsRequest) returns (StatisticsResponse);
    rpc GetUtilization(UtilizationRequest) returns (UtilizationResponse);
    
    // Health check
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ============ МОДЕЛИ ============

message Lesson {
    int32 id = 1;
    
    // Время
    int32 day_of_week = 2;
    int32 time_slot = 3;
    string week_type = 4;                      // 'odd', 'even', 'both'
    string start_time = 5;                     // "09:00"
    string end_time = 6;                       // "10:30"
    
    // Дисциплина
    int32 discipline_id = 7;
    string discipline_name = 8;
    string discipline_code = 9;
    
    // Преподаватель
    int32 teacher_id = 10;
    string teacher_name = 11;
    
    // Группа
    int32 group_id = 12;
    string group_name = 13;
    int32 group_size = 14;
    
    // Аудитория
    int32 classroom_id = 15;
    string classroom_name = 16;
    string building_name = 17;
    
    // Тип
    string lesson_type = 18;
    
    // Семестр
    int32 semester = 19;
    string academic_year = 20;
    
    // Статус
    bool is_active = 21;
    string status = 22;
    
    // Примечания
    string notes = 23;
    
    // Метаданные
    string created_at = 24;
    string updated_at = 25;
}

message ScheduleChange {
    int32 id = 1;
    int32 schedule_id = 2;
    
    string change_type = 3;
    string old_values = 4;                     // JSON
    string new_values = 5;                     // JSON
    
    string change_reason = 6;
    int32 generation_id = 7;
    
    int32 changed_by = 8;
    string changed_by_name = 9;
    string changed_at = 10;
}

message Conflict {
    int32 id = 1;
    
    int32 day_of_week = 2;
    int32 time_slot = 3;
    string week_type = 4;
    
    string conflict_type = 5;
    int32 entity_id = 6;
    string entity_name = 7;
    
    repeated int32 schedule_ids = 8;
    repeated Lesson conflicting_lessons = 9;
    
    string status = 10;
    string severity = 11;
    
    string detected_at = 12;
}

message Snapshot {
    int32 id = 1;
    string name = 2;
    string description = 3;
    
    int32 semester = 4;
    string academic_year = 5;
    
    int32 total_lessons = 6;
    int32 total_teachers = 7;
    int32 total_groups = 8;
    
    int32 fitness_score = 9;
    int32 conflicts_count = 10;
    
    string snapshot_type = 11;
    string created_at = 12;
    string created_by_name = 13;
}

// ============ ЗАПРОСЫ ============

message GroupScheduleRequest {
    int32 group_id = 1;
    
    // Фильтры
    int32 day_of_week = 2;                     // optional
    string week_type = 3;                      // optional
    int32 semester = 4;
    string academic_year = 5;
    
    bool only_active = 6;                      // default: true
}

message TeacherScheduleRequest {
    int32 teacher_id = 1;
    
    int32 day_of_week = 2;
    string week_type = 3;
    int32 semester = 4;
    string academic_year = 5;
    
    bool only_active = 6;
}

message ClassroomScheduleRequest {
    int32 classroom_id = 1;
    
    int32 day_of_week = 2;
    string week_type = 3;
    int32 semester = 4;
    string academic_year = 5;
    
    bool only_active = 6;
}

message DayScheduleRequest {
    int32 day_of_week = 1;
    string week_type = 2;
    int32 semester = 3;
    string academic_year = 4;
    
    // Фильтры
    repeated int32 group_ids = 5;
    repeated int32 teacher_ids = 6;
    repeated int32 classroom_ids = 7;
}

message GetLessonRequest {
    int32 id = 1;
}

message CreateLessonRequest {
    int32 day_of_week = 1;
    int32 time_slot = 2;
    string week_type = 3;
    
    string discipline_name = 4;
    string discipline_code = 5;
    
    int32 teacher_id = 6;
    int32 group_id = 7;
    int32 classroom_id = 8;
    
    string lesson_type = 9;
    int32 semester = 10;
    string academic_year = 11;
    
    string notes = 12;
    int32 created_by = 13;
}

message UpdateLessonRequest {
    int32 id = 1;
    
    // Все поля optional
    int32 day_of_week = 2;
    int32 time_slot = 3;
    string week_type = 4;
    int32 classroom_id = 5;
    string notes = 6;
    
    int32 updated_by = 7;
    string change_reason = 8;
}

message DeleteLessonRequest {
    int32 id = 1;
    int32 deleted_by = 2;
    string reason = 3;
}

message BulkCreateRequest {
    repeated CreateLessonRequest lessons = 1;
    int32 generation_id = 2;                   // Из ms-agent
}

message BulkCreateResponse {
    bool success = 1;
    int32 created_count = 2;
    int32 failed_count = 3;
    repeated string errors = 4;
}

// ============ АКТИВАЦИЯ ============

message ActivateRequest {
    int32 generation_id = 1;                   // ID генерации из ms-agent
    int32 semester = 2;
    string academic_year = 3;
    
    bool deactivate_old = 4;                   // Деактивировать старое
    bool create_snapshot = 5;                  // Создать снимок перед активацией
    
    int32 activated_by = 6;
}

message ActivateResponse {
    bool success = 1;
    int32 activated_count = 2;
    int32 deactivated_count = 3;
    int32 snapshot_id = 4;                     // Если create_snapshot = true
    string message = 5;
}

message DeactivateRequest {
    int32 semester = 1;
    string academic_year = 2;
    int32 deactivated_by = 3;
}

message DeactivateResponse {
    bool success = 1;
    int32 deactivated_count = 2;
}

// ============ КОНФЛИКТЫ ============

message ConflictCheckRequest {
    oneof target {
        int32 lesson_id = 1;
        CreateLessonRequest new_lesson = 2;
    }
}

message ConflictResponse {
    bool has_conflicts = 1;
    repeated Conflict conflicts = 2;
}

message GetConflictsRequest {
    int32 semester = 1;
    string academic_year = 2;
    
    repeated string conflict_types = 3;        // ['teacher', 'group', 'classroom']
    string status = 4;                         // 'active', 'resolved'
}

message ConflictsListResponse {
    repeated Conflict conflicts = 1;
    int32 total_count = 2;
}

message ResolveConflictRequest {
    int32 conflict_id = 1;
    string resolution = 2;
    int32 resolved_by = 3;
}

message ResolveResponse {
    bool success = 1;
    string message = 2;
}

// ============ ИСТОРИЯ ============

message HistoryRequest {
    int32 semester = 1;
    string academic_year = 2;
    
    int32 limit = 3;                           // default: 100
    string change_type = 4;                    // optional filter
    int32 changed_by = 5;                      // optional filter
}

message LessonHistoryRequest {
    int32 lesson_id = 1;
}

message HistoryResponse {
    repeated ScheduleChange changes = 1;
    int32 total_count = 2;
}

// ============ СНИМКИ ============

message SnapshotRequest {
    string name = 1;
    string description = 2;
    
    int32 semester = 3;
    string academic_year = 4;
    
    int32 created_by = 5;
}

message SnapshotResponse {
    Snapshot snapshot = 1;
    string message = 2;
}

message ListSnapshotsRequest {
    int32 semester = 1;
    string academic_year = 2;
    
    string snapshot_type = 3;
}

message SnapshotsListResponse {
    repeated Snapshot snapshots = 1;
    int32 total_count = 2;
}

message RestoreSnapshotRequest {
    int32 snapshot_id = 1;
    bool create_backup = 2;                    // Создать backup текущего
    int32 restored_by = 3;
}

message RestoreResponse {
    bool success = 1;
    int32 restored_count = 2;
    int32 backup_snapshot_id = 3;
    string message = 4;
}

message CompareRequest {
    int32 snapshot1_id = 1;
    int32 snapshot2_id = 2;
}

message CompareResponse {
    int32 added_lessons = 1;
    int32 removed_lessons = 2;
    int32 modified_lessons = 3;
    string differences_json = 4;
}

// ============ ЭКСПОРТ ============

message ExportRequest {
    oneof target {
        int32 group_id = 1;
        int32 teacher_id = 2;
        int32 classroom_id = 3;
        bool all_schedule = 4;
    }
    
    int32 semester = 5;
    string academic_year = 6;
    string format = 7;                         // 'excel', 'pdf', 'ical'
}

message ExportResponse {
    bool success = 1;
    bytes file_data = 2;
    string filename = 3;
    string content_type = 4;
}

// ============ ПОИСК ============

message SearchRequest {
    string query = 1;                          // Поиск по тексту
    
    int32 semester = 2;
    string academic_year = 3;
    
    int32 limit = 4;
}

message ScheduleResponse {
    repeated Lesson lessons = 1;
    int32 total_count = 2;
}

message LessonResponse {
    Lesson lesson = 1;
    string message = 2;
}

message DeleteResponse {
    bool success = 1;
    string message = 2;
}

// ============ СТАТИСТИКА ============

message StatisticsRequest {
    int32 semester = 1;
    string academic_year = 2;
    
    string start_date = 3;                     // optional
    string end_date = 4;                       // optional
}

message StatisticsResponse {
    int32 total_lessons = 1;
    int32 total_hours = 2;
    
    int32 lecture_count = 3;
    int32 practice_count = 4;
    int32 lab_count = 5;
    
    map<int32, int32> lessons_per_day = 6;    // day -> count
    map<string, int32> lessons_per_type = 7;
    
    float avg_fitness_score = 8;
    int32 total_conflicts = 9;
    float preference_satisfaction = 10;
}

message UtilizationRequest {
    int32 semester = 1;
    string academic_year = 2;
    
    string entity_type = 3;                    // 'teacher', 'classroom', 'group'
}

message UtilizationResponse {
    repeated UtilizationItem items = 1;
}

message UtilizationItem {
    int32 entity_id = 1;
    string entity_name = 2;
    float utilization_percentage = 3;
    int32 total_hours = 4;
    int32 busy_slots = 5;
    int32 free_slots = 6;
}

// ============ HEALTH CHECK ============

message HealthCheckRequest {}

message HealthCheckResponse {
    string status = 1;
    string version = 2;
}

